src/optimizer/rules/implementation_rules.cpp

```diff
@@ -428,6 +429,84 @@ void LogicalAggregateToPhysicalAggregate::Transform(common::ManagedPointer<Opera
   transformed->emplace_back(std::move(result));
 }
 
+///////////////////////////////////////////////////////////////////////////////
+/// LogicalInnerJoinToPhysicalInnerIndexJoin
+///////////////////////////////////////////////////////////////////////////////
+LogicalInnerJoinToPhysicalInnerIndexJoin::LogicalInnerJoinToPhysicalInnerIndexJoin() {
+  type_ = RuleType::INNER_JOIN_TO_INDEX_JOIN;
+
+  auto left_child = new Pattern(OpType::LEAF);
+  auto right_child = new Pattern(OpType::LOGICALGET);
+
+  // Initialize a pattern for optimizer to match
+  match_pattern_ = new Pattern(OpType::LOGICALINNERJOIN);
+
+  // Add node - we match join relation R and S
+  match_pattern_->AddChild(left_child);
+  match_pattern_->AddChild(right_child);
+}
+
+bool LogicalInnerJoinToPhysicalInnerIndexJoin::Check(common::ManagedPointer<OperatorNode> plan,
+                                                     OptimizationContext *context) const {
+  (void)context;
+  (void)plan;
+  UNUSED_ATTRIBUTE const auto join = plan->GetOp().As<LogicalInnerJoin>();
+  TERRIER_ASSERT(join != nullptr, "Plan should be a Inner Join");
+  TERRIER_ASSERT(plan->GetChildren().size() == 2, "Inner Join should have two children");
+
+  // Get the "right" inner child
+  auto rchild = plan->GetChildren()[1];
+  UNUSED_ATTRIBUTE auto get = rchild->GetOp().As<LogicalGet>();
+  TERRIER_ASSERT(get != nullptr, "Right child should be a GET");
+
+  // Check whether the right child can be done as an index scan
+  LogicalGetToPhysicalIndexScan idx_scan_check;
+  return idx_scan_check.Check(rchild, context);
```

Comment:
Would this check include the case where the `LogicalGet` of the right child cannot be done as an `IndexScan`, but only when you "push" the join predicate into the right child (i.e., the "`new_child`" in `Transform()`), then you can do an `IndexScan`?

Reply:
Yes. The `Check` function for `LogicalGetToPhysicalIndexScan` does not do predicate checking. It only confirms that there are indexes for the right table.